File: NEsper/NEsper.Common/common/internal/epl/datetime/eval/DatetimeMethodResolver.java
package com.espertech.esper.common.@internal.epl.datetime.eval;

import com.espertech.esper.common.client.hook.datetimemethod.DateTimeMethodDescriptor;
import com.espertech.esper.common.client.hook.datetimemethod.DateTimeMethodForgeFactory;
import com.espertech.esper.common.client.hook.datetimemethod.DateTimeMethodInitializeContext;
import com.espertech.esper.common.@internal.epl.datetime.plugin.DTMPluginForgeFactory;
import com.espertech.esper.common.@internal.epl.expression.core.ExprValidationException;
import com.espertech.esper.common.@internal.settings.ClasspathImportException;
import com.espertech.esper.common.@internal.settings.ClasspathImportServiceCompileTime;
import com.espertech.esper.common.@internal.util.JavaClassHelper;

import java.util.Locale;

public class DatetimeMethodResolver {
    public static boolean isDateTimeMethod(String name, ClasspathImportServiceCompileTime classpathImportService) throws ExprValidationException {
        for (DatetimeMethodBuiltin e : DatetimeMethodBuiltin.values()) {
            if (e.getNameCamel().toLowerCase(Locale.ENGLISH).equals(name.toLowerCase(Locale.ENGLISH))) {
                return true;
            }
        }
        try {
            return classpathImportService.resolveDateTimeMethod(name) != null;
        } catch (ClasspathImportException e) {
            throw new ExprValidationException("Failed to resolve date-time-method '" + name + "': " + e.getMessage(), e);
        }
    }

    public static DatetimeMethodDesc fromName(String name, ClasspathImportServiceCompileTime classpathImportService) throws ExprValidationException {
        for (DatetimeMethodBuiltin e : DatetimeMethodBuiltin.values()) {
            if (e.getNameCamel().toLowerCase(Locale.ENGLISH).equals(name.toLowerCase(Locale.ENGLISH))) {
                return e.getDescriptor();
            }
        }

        try {
            Class factory = classpathImportService.resolveDateTimeMethod(name);
            if (factory != null) {
                DateTimeMethodForgeFactory forgeFactory = (DateTimeMethodForgeFactory) JavaClassHelper.instantiate(DateTimeMethodForgeFactory.class, factory);
                DateTimeMethodDescriptor descriptor = forgeFactory.initialize(new DateTimeMethodInitializeContext());
                DTMPluginForgeFactory plugin = new DTMPluginForgeFactory(forgeFactory);
                return new DatetimeMethodDesc(DatetimeMethodEnum.PLUGIN, plugin, descriptor.getFootprints());
            }
        } catch (Exception ex) {
            throw new ExprValidationException("Failed to resolve date-time-method '" + name + "' :" + ex.getMessage(), ex);
        }
        return null;
    }
}

File: src/NEsper.Common/common/internal/epl/datetime/eval/DatetimeMethodResolver.cs
///////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2006-2024 Esper Team. All rights reserved.                           /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
///////////////////////////////////////////////////////////////////////////////////////

using System;

using com.espertech.esper.common.client.hook.datetimemethod;
using com.espertech.esper.common.@internal.epl.datetime.plugin;
using com.espertech.esper.common.@internal.epl.expression.core;
using com.espertech.esper.common.@internal.settings;
using com.espertech.esper.common.@internal.util;
using com.espertech.esper.compat;

namespace com.espertech.esper.common.@internal.epl.datetime.eval
{
    public class DatetimeMethodResolver
    {
        public static bool IsDateTimeMethod(
            string name,
            ImportServiceCompileTime importService)
        {
            foreach (var e in EnumHelper.GetValues<DatetimeMethodBuiltin>()) {
                var eNameCamel = e.GetNameCamel();
                if (eNameCamel.Equals(name, StringComparison.InvariantCultureIgnoreCase)) {
                    return true;
                }
            }

            try {
                return importService.ResolveDateTimeMethod(name) != null;
            }
            catch (ImportException e) {
                throw new ExprValidationException("Failed to resolve date-time-method '" + name + "': " + e.Message, e);
            }
        }

        public static DatetimeMethodDesc FromName(
            string name,
            ImportServiceCompileTime importService)
        {
            foreach (var e in EnumHelper.GetValues<DatetimeMethodBuiltin>()) {
                var eNameCamel = e.GetNameCamel();
                if (eNameCamel.Equals(name, StringComparison.InvariantCultureIgnoreCase)) {
                    return e.GetDescriptor();
                }
            }

            try {
                var factory = importService.ResolveDateTimeMethod(name);
                if (factory != null) {
                    var forgeFactory = TypeHelper.Instantiate<DateTimeMethodForgeFactory>(factory);
                    var descriptor = forgeFactory.Initialize(new DateTimeMethodInitializeContext());
                    var plugin = new DTMPluginForgeFactory(forgeFactory);
                    return new DatetimeMethodDesc(DateTimeMethodEnum.PLUGIN, plugin, descriptor.Footprints);
                }
            }
            catch (Exception ex) {
                throw new ExprValidationException(
                    "Failed to resolve date-time-method '" + name + "' :" + ex.Message,
                    ex);
            }

            return null;
        }
    }
} // end of namespace